src/nv/storage/index.rs

```diff
@@ -61,47 +61,362 @@ bitfield! {
     pub struct NvIndexAttributes(TPMA_NV);
     impl Debug;
     // NV Index Attributes
-    pub pp_write, set_pp_write: 0;
-    pub owner_write, set_owner_write: 1;
-    pub auth_write, set_auth_write: 2;
-    pub policy_write, set_policy_write: 3;
+    pub pp_write, _: 0;
+    _, set_pp_write: 0;
+    pub owner_write, _: 1;
+    _, set_owner_write: 1;
+    pub auth_write, _: 2;
+    _, set_auth_write: 2;
+    pub policy_write, _: 3;
+    _, set_policy_write: 3;
     TPM2_NT, tss_index_type, _: 7, 4; // private getter
-    pub TPM2_NT, from into NvIndexType, _, set_index_type: 7, 4;
+    TPM2_NT, from into NvIndexType, _, set_index_type: 7, 4;
     // Reserved 9,8
-    pub policy_delete, set_policy_delete: 10;
-    pub write_locked, set_write_locked: 11;
-    pub write_all, set_write_all: 12;
-    pub write_define, set_write_define: 13;
-    pub write_stclear, set_write_stclear: 14;
-    pub global_lock, set_global_lock: 15;
-    pub pp_read, set_pp_read: 16;
-    pub owner_read, set_owner_read: 17;
-    pub auth_read, set_auth_read: 18;
-    pub policy_read, set_policy_read: 19;
+    pub policy_delete, _: 10;
+    _, set_policy_delete: 10;
+    pub write_locked, _: 11;
+    _, set_write_locked: 11;
+    pub write_all, _: 12;
+    _, set_write_all: 12;
+    pub write_define, _: 13;
+    _, set_write_define: 13;
+    pub write_stclear, _: 14;
+    _, set_write_stclear: 14;
+    pub global_lock, _: 15;
+    _, set_global_lock: 15;
+    pub pp_read, _: 16;
+    _, set_pp_read: 16;
+    pub owner_read, _: 17;
+    _, set_owner_read: 17;
+    pub auth_read, _: 18;
+    _, set_auth_read: 18;
+    pub policy_read, _: 19;
+    _, set_policy_read: 19;
     // Reserved 24, 20
-    pub no_da, set_no_da: 25;
-    pub orderly, set_orderly: 26;
-    pub clear_stclear, set_clear_stclear: 27;
-    pub read_locked, set_read_locked: 28;
-    pub written, set_written: 29;
-    pub platform_create, set_platform_create: 30;
-    pub read_stclear, set_read_stclear: 31;
+    pub no_da, _: 25;
+    _, set_no_da: 25;
+    pub orderly, _: 26;
+    _, set_orderly: 26;
+    pub clear_stclear, _: 27;
+    _, set_clear_stclear: 27;
+    pub read_locked, _: 28;
+    _, set_read_locked: 28;
+    pub written, _: 29;
+    _, set_written: 29;
+    pub platform_create, _: 30;
+    _, set_platform_create: 30;
+    pub read_stclear, _: 31;
+    _, set_read_stclear: 31;
 }
 
 impl NvIndexAttributes {
+    /// Returns the `NvIndexType` of the `NvIndexAttributes`
     pub fn index_type(&self) -> Result<NvIndexType> {
         NvIndexType::try_from(self.tss_index_type())
     }
+
+    /// Performs checks on the `NvIndexAttributes` and returns
+    /// `Ok(self)` if it passes all checks.
+    ///
+    /// # Errors
+    /// Returns an error if some attributes are missing
+    /// or are in conflict with eachother.
+    pub fn as_validated(&self) -> Result<NvIndexAttributes> {
```

Comment:
Is it possible to keep this as `validate(&self) -> Result<()>` and in the builder deconstruct it to return the `NvIndexAttributes` already found inside it?

Reply:
fixed t

Reply:
I think it is the way to go. Because as_validated would have been nice if ```&mut self``` had been required to perform the checks or if it was not possible to copy ```self``` trivially.